### PR TITLE
Bump Python to 3.9

### PR DESCRIPTION
**Pull request by behrmann**
With 3.7 we are currently at the lower end of supported Python versions. The current Debian stable, that will become oldstable next year, runs 3.9, so it's maybe time for us to get the goodness like `Path.readlink`, `missing_ok=` on `Path`s, `str.removesuffix` and `str.removeprefix` as well as much fewer typing imports.

This is based on #1280, so the first commit can be ignored. 

@DaanDeMeyer Since this PR, too, might generate a lot of merge conflicts for you: The last commit can be mostly ignored (i.e. if it makes it easier dropped and postponed), it was generated with [pyupgrade](https://github.com/asottile/pyupgrade) and the remaining imports were found by pyflakes (which we already run in CI).

### CODE REVIEW

**Comment by behrmann** [comment]
Added the "dont merge" because this contains #1280, which should be discussed there.

**Comment by DaanDeMeyer** [comment]
Yeah this one can wait until after repart, fixing merge conflicts for the other PR was painful enough already

**Comment by behrmann** [comment]
I feel your pain :) But no worries, this one is extremely easy to recreate, so no hurry.

**Comment by keszybz** [comment]
> ukify requires python 3.11 I think? So we can bump even more

Does it? At least 3.10 seems to accept the syntax without issues.

**Comment by behrmann** [comment]
Yeah, 3.10 should be sufficient and I wanted to ask about that too, but I haven't found the time to rebase this yet, but should tomorrow.

Bumping to 3.11 seems too early to me, since 3.11 hasn't landed in Arch and Debian yet. I'm not too optimistic that 3.10 will make it into Debian bookworm; they're trying, though. It's one thing we ask people to build systemd, but that's quite easy, but asking people to manage a separate Python toolchain might be asking too much.

Bumping to 3.10 would be a possibility and we basically already require it because of ukify, but I'm a bit hesitant. ukify is pretty easy to backport to 3.9, but once we move mkosi proper to 3.10 we're back to requiring people to build a separate cpython on distros that don't offer it (Arch and Debian stable).

**Comment by DaanDeMeyer** [comment]
Hmm, if bookworm doesn't get 3.10 we'd probably do good to downgrade ukify's required version to 3.9

**Comment by behrmann** [comment]
Bookworm *will* get 3.10, but it's - at least to me - unclear whether it will get 3.11.

**Comment by keszybz** [comment]
Yeah, that was my plan when writing ukify: I wanted to start with the modern syntax and other frills because it's nicer this way, but I have no qualms about adding workarounds for older versions. Supporting 3.9 would require getting rid of `match`, which I think works quite well in that code, but support for older distros is more important.

**Review by DaanDeMeyer — APPROVED** [review]

**Pull request merged**
